torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  // If cpu_end_time is not recorded in this iteration,
+  // avg_time will return invalid value.
+  // For some cases like DDP runs on non-sync mode, backward compute
+  // end time can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  long num_iters = reducer_->num_iterations_;
+  avg_time = ((cpu_end_time - cpu_start_time) + avg_time * (num_iters - 1)) /
+      num_iters;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    at::cuda::CUDAEvent& gpu_start,
+    at::cuda::CUDAEvent& gpu_end) {
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  long num_iters = reducer_->num_iterations_;
+  float milliseconds = gpu_start.elapsed_time(gpu_end);
+  // If gpu_end is not recorded in this iteration,
+  // milliseconds will have invalid value.
+  // For some cases like DDP runs on non-sync mode,
+  // gpu_end can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
+              avg_time * (num_iters - 1)) /
+      num_iters;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
```

Comment:
Not sure if it's the best pattern to rely on another class's mutex to access it's private attributes, but I suppose it's probably fine since it's a friend class.

Reply:
that is also what I thought, logger depends on reducer's data, they needs to be synced up.